packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -1884,6 +1861,476 @@ class OwnerValidGadget : public GadgetT
+    }
+
+    void generate_r1cs_constraints()
+    {
+        notSignB.generate_r1cs_constraints();
+        signedAddGadget.generate_r1cs_constraints();
+    }
+
+    const SignedVariableT result() const
+    {
+        return signedAddGadget.result();
+    }
+};
+
+// sA * sB / C
+// Always round down. For negative values we round away from zero.
+// So if the result is negative and the remainder of the division is non-zero we add 1 to abs(result).
+class SignedMulDivGadget : public GadgetT
+{
+  public:
+    MulDivGadget res;
+    EqualGadget sign_math;
+
+    EqualGadget remainderZero;
+    AndGadget bias;
+
+    AddGadget value;
+
+    EqualGadget isZero;
+    TernaryGadget sign;
+
+    SignedMulDivGadget(
+      ProtoboardT &pb,
+      const Constants &constants,
+      const SignedVariableT &_value,
+      const SignedVariableT &_numerator,
+      const VariableT &_denominator,
+      unsigned int numBitsValue,
+      unsigned int numBitsNumerator,
+      unsigned int numBitsDenominator,
+      const std::string &prefix)
+        : GadgetT(pb, prefix),
+
+          res(
+            pb,
+            constants,
+            _value.value,
+            _numerator.value,
+            _denominator,
+            numBitsValue,
+            numBitsNumerator,
+            numBitsDenominator,
+            FMT(prefix, ".res")),
+          sign_math(pb, _value.sign, _numerator.sign, FMT(prefix, ".sign")),
+
+          remainderZero(pb, res.getRemainder(), constants._0, FMT(prefix, ".remainderZero")),
+          bias(pb, {remainderZero.result(), sign_math.result()}, FMT(prefix, ".bias")),
+
+          value(
+            pb,
+            res.result(),
+            /*bias.result()*/ constants._0,
+            numBitsValue + numBitsNumerator - numBitsDenominator,
+            FMT(prefix, ".value")),
+
+          isZero(pb, value.result(), constants._0, FMT(prefix, ".isZero")),
+          sign(pb, isZero.result(), constants._0, sign_math.result(), FMT(prefix, ".sign"))
+    {
+    }
+
+    void generate_r1cs_witness()
+    {
+        res.generate_r1cs_witness();
+        sign_math.generate_r1cs_witness();
+
+        remainderZero.generate_r1cs_witness();
+        bias.generate_r1cs_witness();
+
+        value.generate_r1cs_witness();
+
+        isZero.generate_r1cs_witness();
+        sign.generate_r1cs_witness();
+    }
+
+    void generate_r1cs_constraints()
+    {
+        res.generate_r1cs_constraints();
+        sign_math.generate_r1cs_constraints();
+
+        remainderZero.generate_r1cs_constraints();
+        bias.generate_r1cs_constraints();
+
+        value.generate_r1cs_constraints();
+
+        isZero.generate_r1cs_constraints();
+        sign.generate_r1cs_constraints();
+    }
+
+    const SignedVariableT result() const
+    {
+        return SignedVariableT(sign.result(), value.result());
+    }
+};
+
+// Calculates [0, 1]**[0, inf[ using an approximation. Best accuracy is achieved around 1**y.
```

Comment:
`[0, 1]**[0, inf]` typo?

Do you mean calculating power(a, b), where 0<=a<=1 and b >=0?

Reply:
what's `1**y`?

Reply:
> [0, 1]**[0, inf] typo?

Infinity is not part of the range because it's impossible, so the bracket is open to show that, but not 100% sure if that's the correct notation in this case.

> Do you mean calculating power(a, b), where 0<=a<=1 and b >=0?

Yes.

> what's 1**y?

`power(1, y)`. Just trying to say that the closer the base is to 1, the better the accuracy.

Reply:
`[0, inf[` =>`[0, inf)`

Reply:
>Just trying to say that the closer the base is to 1, the better the accuracy.
Maybe just say the above in the sourcefile.

Reply:
> [0, inf[ =>[0, inf)

Didn't know a parenthesis could also be used for this, both are standard notation it seems: https://en.wikipedia.org/wiki/Interval_(mathematics)#Including_or_excluding_endpoints